### PR TITLE
feat(toolbars): use toolbar base component in toolbar plugins

### PR DESCRIPTION
**Pull request by betterdancing**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

## Release Notes

- **New Features**
	- Introduced `ToolbarBaseButton` component, allowing for customizable button features within toolbars.
	- Enhanced toolbar components with new `options` prop for improved configurability.

- **Bug Fixes**
	- Improved rendering logic for various toolbar components to ensure accurate display based on state and props.

- **Refactor**
	- Streamlined toolbar component structures, consolidating functionalities into the new `toolbar-base` component for better organization and maintainability.
	- Refactored multiple toolbar components to utilize the `toolbar-base`, enhancing usability and clarity.

- **Documentation**
	- Updated metadata for toolbar configurations to improve clarity and extensibility.

### CODE REVIEW

File: packages/toolbars/breadcrumb/package.json [hunk 1]
```diff
@@ -27,7 +27,8 @@
     "@opentiny/tiny-engine-common": "workspace:*",
     "@opentiny/tiny-engine-http": "workspace:*",
     "@opentiny/tiny-engine-meta-register": "workspace:*",
-    "@opentiny/tiny-engine-utils": "workspace:*"
+    "@opentiny/tiny-engine-utils": "workspace:*",
+    "@opentiny/tiny-engine-toolbar-checkinout": "workspace:*"
```
Comment (hexqi) [inline packages/toolbars/breadcrumb/package.json]: 这个依赖已经没有了，需要删掉

File: packages/layout/src/toolbarBuiltIn/toolbarButton.vue
Comment (hexqi) [inline packages/layout/src/toolbarBuiltIn/toolbarButton.vue:1]: 文件夹名尽量用小写，中横线连字符；
Vue组件文件名首字母要大写(类似于Class的规范)

File: packages/layout/src/toolbarBuiltIn/toolbarButton.vue [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      type: String,
+      default: ''
+    },
+    content: {
+      type: String,
+      default: ''
+    },
+    options: {
+      type: Object,
+      default: () => {}
```
Comment (hexqi) [inline packages/layout/src/toolbarBuiltIn/toolbarButton.vue]: 少了括号， ==> `() => ({})`

File: packages/layout/src/toolbarBuiltIn/toolbarIcon.vue [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    :open-delay="1000"
+    popper-class="toolbar-right-popover"
+    append-to-body
+    :content="content"
+  >
+    <template #reference>
+      <span class="icon">
+        <span :class="[options?.useDefaultClass ? 'icon-hides' : '']">
+          <svg-icon :name="icon"></svg-icon>
+          <span v-if="!isSaved() && options?.showDots" class="dots"></span>
```
Comment (hexqi) [inline packages/layout/src/toolbarBuiltIn/toolbarIcon.vue]: ToolbarIcon是一个基础组件，不应该依赖useCanvas().isSaved

File: packages/layout/src/ToolbarBaseComponent.vue [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  emits: ['click-api'],
+  setup(props, { emit }) {
+    const state = reactive({
+      icon: computed(() => props.icon),
+      content: computed(() => props.content),
+      options: computed(() => props.options)
+    })
+    
+    const handelClick = () => {
+      emit('click-api')
```
Comment (hexqi) [inline packages/layout/src/ToolbarBaseComponent.vue]: 事件名建议就叫click

File: packages/layout/src/toolbarBuiltIn/toolbarCustom.vue
Comment (hexqi) [inline packages/layout/src/toolbarBuiltIn/toolbarCustom.vue:1]: toolbarCustom这个组件是不是没有用到

File: packages/toolbars/breadcrumb/meta.js [hunk 5]
```diff
@@ -2,6 +2,10 @@ export default {
   id: 'engine.toolbars.breadcrumb',
   type: 'toolbars',
   title: 'breadcrumb',
-  icon: '',
-  align: 'left'
+  icon: {
+    default: ''
+  },
+  align: 'left',
+  renderType: 'slot',
```
Comment (hexqi) [inline packages/toolbars/breadcrumb/meta.js]: renderType也放到options里面吧，协议里外层没有这个字段

File: packages/layout/src/toolbarBuiltIn/toolbarButton.vue [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    v-bind="options.props"
+    :style="options.style"
+    :class="[options?.useDefaultClass ? 'toolbar-button' : '']"
+  >
+    <span class="svg-wrap">
+      <svg-icon :name="icon"></svg-icon>
+      <span v-if="!isSaved() && options?.showDots" class="dots"></span>
+    </span>
+    <span class="save-title">{{ content }}</span>
+    <slot name="button-extends"></slot>
```
Comment (yy-wow) [inline packages/layout/src/toolbarBuiltIn/toolbarButton.vue]: 只有一个插槽，且含义与组件本身相近，建议直接使用默认插槽

File: packages/layout/src/ToolbarBaseComponent.vue [hunk 7]
```diff
@@ -0,0 +1,85 @@
+<template>
+  <span class="toolbar-item-wrap" @click="handelClick">
+    <component :is="getRender()" v-bind="state">
+      <template #button-extends>
+        <slot name="button-extends"></slot>
```
Comment (yy-wow) [inline packages/layout/src/ToolbarBaseComponent.vue]: 这里含义是传递一个按钮组件吗？插槽名建议直接叫"button"

File: packages/layout/src/ToolbarBaseComponent.vue [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      </template>
+    </component>
+    <slot name="extends"></slot>
```
Comment (yy-wow) [inline packages/layout/src/ToolbarBaseComponent.vue]: extends和插槽本身含义相近，可以直接使用默认插槽

**Review by hexqi — COMMENTED** [review]
分支有冲突

File: packages/layout/index.js [hunk 9]
```diff
@@ -30,4 +31,4 @@ export default {
   metas: [LayoutService]
 }
 
-export { LayoutService }
+export { ToolbarBaseComponent, LayoutService }
```
Comment (hexqi) [inline packages/layout/index.js]: 其他插件是不应该依赖layout这个包的，这个组件得放到common中

File: packages/design-core/index.js [hunk 10]
```diff
@@ -30,7 +30,7 @@ export { default as Robot } from '@opentiny/tiny-engine-plugin-robot'
 export { default as Props, PropertiesService, PropertyService } from '@opentiny/tiny-engine-setting-props'
 export { default as Events } from '@opentiny/tiny-engine-setting-events'
 export { default as Styles } from '@opentiny/tiny-engine-setting-styles'
-export { default as Layout, LayoutService } from '@opentiny/tiny-engine-layout'
+export { default as Layout, ToolbarBaseComponent, LayoutService } from '@opentiny/tiny-engine-layout'
```
Comment (hexqi) [inline packages/design-core/index.js]: ToolbarBaseComponent这个名字不好，用ToolbarBase吧

File: packages/toolbars/breadcrumb/package.json [hunk 11]
```diff
@@ -27,7 +27,8 @@
     "@opentiny/tiny-engine-common": "workspace:*",
     "@opentiny/tiny-engine-http": "workspace:*",
     "@opentiny/tiny-engine-meta-register": "workspace:*",
-    "@opentiny/tiny-engine-utils": "workspace:*"
+    "@opentiny/tiny-engine-utils": "workspace:*",
+    "@opentiny/tiny-engine-layout": "workspace:*"
```
Comment (hexqi) [inline packages/toolbars/breadcrumb/package.json]: 插件不应该依赖layout包

File: packages/layout/src/ToolbarBaseComponent.vue
Comment (chilingling) [inline packages/layout/src/ToolbarBaseComponent.vue:1]: @coderabbitai Please give this component a few semanticized component names based on what this component does.

File: packages/common/component/index.js [hunk 12]
```diff
@@ -53,6 +53,7 @@ export { default as SplitPanes } from './SplitPanes.vue'
 export { default as Pane } from './Pane.vue'
 export { default as I18nInput } from './I18nInput.vue'
 export { default as CanvasDragItem } from './CanvasDragItem.vue'
+export { default as ToolbarBase } from './ToolbarBase.vue'
```
Comment (hexqi) [inline packages/common/component/index.js:56]: design-core的index里也要导出下：`export { ToolbarBase } from '@opentiny/tiny-engine-common'`

File: packages/toolbars/breadcrumb/src/Main.vue [hunk 13]
```diff
@@ -1,39 +1,54 @@
...
+          </tiny-breadcrumb>
+        </div>
+
+        <tiny-button
+          class="publish"
+          v-if="breadcrumbData[0] === CONSTANTS.BLOCKTEXT"
+          @click="publishBlock()"
+          type="primary"
+          size="small"
+          >发布区块</tiny-button
```
Comment (hexqi) [inline packages/toolbars/breadcrumb/src/Main.vue]: tiny-button的标签可以跟上面对齐下

File: packages/toolbars/breadcrumb/src/Main.vue [hunk 14]
```diff
@@ -1,39 +1,54 @@
...
     TinyBreadcrumbItem: BreadcrumbItem,
     BlockDeployDialog,
-    TinyButton: Button
+    TinyButton: Button,
+    ToolbarBase
+  },
+  props: {
+    options: {
+      type: Object,
+      default: () => {}
```
Comment (hexqi) [inline packages/toolbars/breadcrumb/src/Main.vue]: 少了括号：`() => ({})`

File: packages/toolbars/clean/meta.js [hunk 15]
```diff
@@ -2,5 +2,10 @@ export default {
   id: 'engine.toolbars.clean',
   type: 'toolbars',
   title: 'clean',
-  icon: ''
+  options: {
+    icon: {
+      default: 'clear'
```
Comment (hexqi) [inline packages/toolbars/clean/meta.js:7]: 单个图标的话可以也支持下之前字符串的格式，`icon: 'xxxx'`

**Review by hexqi — DISMISSED** [review]

**Comment by hexqi** [comment]
image
image
按钮的样式有点问题，多了一个边框

**Review by hexqi — APPROVED** [review]

**Pull request merged**